### PR TITLE
Replace BSGJSONSerialization class with functions

### PR DESCRIPTION
**Pull request by nickdowell**
## Goal

Simplify JSON (de)serialization code and reduce binary size.

## Changeset

Replaces the `BSGJSONSerialization` class and static methods with equivalent functions.

Deserializes JSON to `NSDictionary` to reduce the number of `[object isKindOfClass:[NSDictinoary class]]` checks throughout the codebase.

Removes unused `NSStream` based methods.

## Testing

Passes all unit and E2E tests.

### CODE REVIEW

File: Bugsnag/Helpers/BSGJSONSerialization.m [hunk 1]
```diff
@@ -7,86 +7,72 @@
...
     }
-}
-
-+ (nullable id)JSONObjectWithData:(NSData *)data options:(NSJSONReadingOptions)opt error:(NSError * __autoreleasing *)error {
     @try {
-        return [NSJSONSerialization JSONObjectWithData:data options:opt error:error];
-    } @catch (NSException *exception) {
-        if (error) {
-            *error = wrapException(exception);
+        if (![NSJSONSerialization isValidJSONObject:(id _Nonnull)obj]) {
```
Comment (kstenerud) [inline Bugsnag/Helpers/BSGJSONSerialization.m:30]: This stuff feels like a repeat of `BSGJSONDictionaryIsValid`

File: Bugsnag/Client/BugsnagClient.m [hunk 2]
```diff
@@ -227,7 +227,7 @@ - (void)start {
     [self computeDidCrashLastLaunch];
 
     // These files can only be overwritten once the previous contents have been read; see -generateEventForLastLaunchWithError:
-    [BSGJSONSerialization writeJSONObject:self.configuration.dictionaryRepresentation toFile:BSGFileLocations.current.configuration options:0 error:nil];
+    BSGJSONWriteDictionaryToFile(self.configuration.dictionaryRepresentation, BSGFileLocations.current.configuration, nil);
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Client/BugsnagClient.m:230]: Yes, there's nothing that could be done in response to failure.

File: Bugsnag/BugsnagSystemState.m [hunk 3]
```diff
@@ -173,9 +173,9 @@ - (void)mutateLaunchState:(nonnull void (^)(NSMutableDictionary *state))block {
 
 - (void)sync {
     NSDictionary *state = self.currentLaunchState;
-    NSAssert([BSGJSONSerialization isValidJSONObject:state], @"BugsnagSystemState cannot be converted to JSON data");
+    NSAssert(BSGJSONDictionaryIsValid(state), @"BugsnagSystemState cannot be converted to JSON data");
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/BugsnagSystemState.m:176]: `NSAssert` is compiled out of non-Debug builds so would not crash a shipping app.

That said, this check isn't really necessary.

File: Bugsnag/Helpers/BSGJSONSerialization.m [hunk 4]
```diff
@@ -7,86 +7,72 @@
...
-+ (nullable NSData *)dataWithJSONObject:(id)obj options:(NSJSONWritingOptions)opt error:(NSError * __autoreleasing *)error {
-    @try {
-        return [NSJSONSerialization dataWithJSONObject:obj options:opt error:error];
-    } @catch (NSException *exception) {
-        if (error) {
-            *error = wrapException(exception);
-        }
+NSData *_Nullable BSGJSONDataFromDictionary(NSDictionary *_Nullable obj, NSError **error) {
+    if (!obj) {
         return nil;
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Helpers/BSGJSONSerialization.m:27]: No - passing a nil dictionary is not an error.

File: Bugsnag/Client/BugsnagClient.m [hunk 5]
```diff
@@ -1043,7 +1043,7 @@ - (void)appHangEnded {
 
 - (nullable BugsnagEvent *)loadAppHangEvent {
     NSError *error = nil;
-    NSDictionary *json = [BSGJSONSerialization JSONObjectWithContentsOfFile:BSGFileLocations.current.appHangEvent options:0 error:&error];
+    NSDictionary *json = BSGJSONDictionaryFromFile(BSGFileLocations.current.appHangEvent, 0, &error);
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Client/BugsnagClient.m:1046]: That would unnecessarily increase the size of this PR.

File: Bugsnag/Helpers/BSGJSONSerialization.m [hunk 6]
```diff
@@ -7,86 +7,72 @@
...
+    return nil;
 }
 
-+ (BOOL)writeJSONObject:(id)JSONObject toFile:(NSString *)file options:(NSJSONWritingOptions)options error:(NSError * __autoreleasing *)errorPtr {
-    if (![BSGJSONSerialization isValidJSONObject:JSONObject]) {
+BOOL BSGJSONWriteDictionaryToFile(NSDictionary *_Nullable JSONObject, NSString *file, NSError **errorPtr) {
+    if (!BSGJSONDictionaryIsValid(JSONObject)) {
         if (errorPtr) {
             *errorPtr = [NSError errorWithDomain:@"BSGJSONSerializationErrorDomain" code:0 userInfo:@{
                 NSLocalizedDescriptionKey: @"Not a valid JSON object"}];
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Helpers/BSGJSONSerialization.m]: Done

File: Bugsnag/Client/BugsnagClient.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Client/BugsnagClient.m:1046]: Also I think `json` is a suitable name for this - it's the argument that will be passed to `-[BugsnagEvent initWithJson:]`

**Review by kstenerud — APPROVED** [review]

**Pull request merged**